angr/exploration_techniques/loop_seer.py

```diff
@@ -0,0 +1,141 @@
+from collections import defaultdict
+import angr
+from . import ExplorationTechnique
+
+class LoopSeer(ExplorationTechnique):
+    """
+    This exploration technique monitors exploration and maintains all
+    loop-related data (well, currently it is just the loop trip counts, but feel
+    free to add something else).
+    """
+
+    def __init__(self, cfg=None, functions=None, loops=None, bound=None, bound_reached=None, discard_stash='spinning'):
+        """
+        :param cfg          : Normalized CFG is required.
+        :param functions    : Function(s) containing the loop(s) to be analyzed.
+        :param loops        : Loop(s) to be analyzed.
+        :param bound        : Limit the number of iteration a loop may be executed.
+        :param bound_reached: If provided, should be a function that takes a SimulationManager and returns
+                              a SimulationManager. Will be called when loop execution reach the given bound.
+                              Default to moving states that exceed the loop limit to a discard stash.
+        :param discard_stash: Name of the stash containing states exceeding the loop limit.
+        """
+
+        super(LoopSeer, self).__init__()
+        self.cfg = cfg
+        self.functions = functions
+        self.bound = bound
+        self.bound_reached = bound_reached
+        self.discard_stash = discard_stash
+
+        self.loops = {}
+        self.current_loop = []
+
+        if type(loops) == angr.analyses.loopfinder.Loop:
+            loops = [loops]
+        
+        if type(loops) in (list, tuple) and all(type(l) == angr.analyses.loopfinder.Loop for l in loops):
+            for loop in loops:
+                print loop.entry
```

Comment:
I'd use debug logging instead of printing this out.

Reply:
Oops, sorry, I forget to remove this...